src/grammar.coffee

```diff
@@ -47,9 +47,14 @@ o = (patternString, action, options) ->
     getAddDataToNodeFunctionString = (first, last) ->
       "yy.addDataToNode(yy, @#{first}#{if last then ", @#{last}" else ''})"
 
+    returnsLoc = /^LOC/.test action
     action = action.replace /LOC\(([0-9]*)\)/g, getAddDataToNodeFunctionString('$1')
     action = action.replace /LOC\(([0-9]*),\s*([0-9]*)\)/g, getAddDataToNodeFunctionString('$1', '$2')
-    performActionFunctionString = "$$ = #{getAddDataToNodeFunctionString(1, patternCount)}(#{action});"
+    performActionFunctionString =
+      if returnsLoc
+        "$$ = #{action};"
+      else
+        "$$ = #{getAddDataToNodeFunctionString(1, patternCount)}(#{action});"
```

Comment:
Do you mind please explaining this change? Why was this necessary?

Reply:
This is part of the `LOC()`-wrapping logic described in the comment above (not sure if you started this review before I'd posted the comments)

Here, we don't do the typical wrapping in `getAddDataToNodeFunctionString()` (ie `LOC()`) if the grammar rule already calls `LOC()` on its return value

Reply:
Yes our comments passed each other in flight. So you’re saying that `getAddDataToNodeFunctionString` isn’t necessary for rules that invoke `LOC`, since `LOC` already adds the missing location data? I _think_ that could be correct, if all that `getAddDataToNodeFunctionString` did was add location data, but doesn’t it also add comments data?